### PR TITLE
luci-app-smartdns: add new package

### PR DESCRIPTION
**Pull request by pymumu**
This luci app supports smartdns.

Depends on openwrt/packages#10797.

Signed-off-by: Nick Peng <pymumu@gmail.com>

### CODE REVIEW

File: applications/luci-app-smartdns/luasrc/model/smartdns.lua [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+-- GNU General Public License for more details.
+--
+-- You should have received a copy of the GNU General Public License
+-- along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+require ("nixio.fs")
+require ("luci.http")
+require ("luci.dispatcher")
+require ("nixio.fs")
+
```
Comment (feckert) [inline applications/luci-app-smartdns/luasrc/model/smartdns.lua]: Please use `local <variable> = require("xxx")`

Comment (pymumu) [reply applications/luci-app-smartdns/luasrc/model/smartdns.lua]: done, plaese review.

**Comment by aparcar** [comment]
Please rewrite this app in [JavaScript instead of Lua](https://github.com/openwrt/luci/issues/3378). If you look at the commit history for other recently added apps they show required files. It is pretty straight forward to convert your existing Lua code to JavaScript.

**Comment by pymumu** [comment]
@aparcar @feckert 
Encountered two problems while rewriting, please help.
1. Load external template html page, used to display service status, don't know how to rewrite.
```javascript
m.section(SimpleSection).template = "smartdns/smartdns_status";
```

2. The following code cannot read or write file(file exists with right permission), and prompts for invalid values ​​when entering some text after saving.
```javascript
o = s.taboption("custom", form.TextValue, "_tmpl",_(""),_("smartdns custom settings"));
o.rows = 20;
o.cfgvalue = function(section_id) {
    return fs.trimmed('/etc/smartdns/custom.conf');
};
o.write = function (section_id, formvalue) {
   return fs.write('/etc/smartdns/custom.conf', formvalue.trim().replace(/\r\n/g, '\n') + '\n');
};
```

[luci-app-smartdns.zip](https://github.com/openwrt/luci/files/4134182/luci-app-smartdns.zip)

**Comment by ysc3839** [comment]
@pymumu 
1.https://github.com/openwrt/luci/blob/master/applications/luci-app-ttyd/htdocs/luci-static/resources/view/ttyd/term.js
2.https://github.com/openwrt/luci/blob/1d5a2fb0b98b5bf9d537843237a5241ada2316ed/applications/luci-app-ksmbd/htdocs/luci-static/resources/view/ksmbd.js#L47-L52
https://github.com/openwrt/luci/blob/master/applications/luci-app-ksmbd/root/usr/share/rpcd/acl.d/luci-app-ksmbd.json

**Comment by pymumu** [comment]
@ysc3839 Problem 2 is solved, thanks.
but I think problem 1 is not the same as the example.

About problem 1, the feature is to display servie status, the status check function calls some shell commands (such as pidof, iptable) periodically and display the result.
for now, I can do is to make the page display some text periodically, but cannot get the results of these commands, because fs.exec is a async function.

the code is like this, dont't knonw how to write smartdnsServiceStatus()  
Not familiar with javascript, request for help.

```javascript
function smartdnsServiceStatus() {
        // how to get the following commands result with fs.exec() ?.
        // pidof smartdns >/dev/null  
        // iptables -t nat -nL PREROUTING 2>/dev/null | grep REDIRECT | grep dpt:53 | grep %q >/dev/null 2>&1  
        // uci get dhcp.@dnsmasq[0].server  
	return "return service status";
}

// section render
s.render = function (section_id) {
	L.Poll.add(function () {
		return L.resolveDefault(smartdnsServiceStatus()).then(function (res) {
			var content = res;
			var view = document.getElementById("service_status");
			view.innerHTML = res;
		});
	});

	return E('div', { class: 'cbi-map' },
		E('div', { class: 'cbi-section' }, [
			E('div', {id: 'service_status' },
				_('Collecting data ...'))
		])
	);
}
```

**Comment by feckert** [comment]
Maybe we can implement an RPC call in the ubus?

**Comment by jow-** [comment]
```javascript
function smartdnsServiceStatus() {
        return Promise.all([
                fs.exec_direct('/bin/pidof', ['smartdns']),
                fs.exec_direct('/usr/sbin/iptables', ['-t', 'nat', '-n', '-L', 'PREROUTING']),
                uci.get_first('dhcp', 'dnsmasq', 'server')
        ]);
}

// section render
s.render = function (section_id) {
	L.Poll.add(function () {
		return L.resolveDefault(smartdnsServiceStatus()).then(function (res) {
			var pid = +res[0],
			    ipt = res[1],
			    srv = res[2];

			var matching_rules = (ipt || '').split(/\n/).filter(function(rule) {
			    return rule.match(/REDIRECT/) && rule.match(/dpt:53/);
			});

			var view = document.getElementById("service_status");
			view.innerHTML = _('PID of service is %s<br />Firewall rule is %s<br />DNS Server is %s').format(
			    isNaN(pid) ? '?' : pid, 
			    matching_rules.length ? 'present' : 'not present',
			    srv);
		});
	});

	return E('div', { class: 'cbi-map' },
		E('div', { class: 'cbi-section' }, [
			E('div', {id: 'service_status' },
				_('Collecting data ...'))
		])
	);
}
```

You need these ACL rules for the above code to work:

```json
{
	"luci-app-smartdns": {
		"description": "...",
		"read": {
			"cgi-io": [ "exec" ],
			"file": {
				"/usr/sbin/iptables -t nat -n -L PREROUTING": [ "exec" ],
				"/bin/pidof smartdns": [ "exec" ]
			}
		}
	}
}
```

**Comment by ysc3839** [comment]
smartdns is using procd to manage process. https://github.com/pymumu/smartdns/blob/744a5409e0fb65b98a8eeb09cdc48bacc0f2a10f/package/openwrt/files/etc/init.d/smartdns#L20
So `pidof` can be replaced with ubus service list.
https://github.com/openwrt/luci/blob/1d5a2fb0b98b5bf9d537843237a5241ada2316ed/applications/luci-app-shadowsocks-libev/htdocs/luci-static/resources/view/shadowsocks-libev/instances.js#L12-L17 https://github.com/openwrt/luci/blob/1d5a2fb0b98b5bf9d537843237a5241ada2316ed/applications/luci-app-shadowsocks-libev/htdocs/luci-static/resources/view/shadowsocks-libev/instances.js#L140-L147

**Comment by pymumu** [comment]
@jow- thanks, problem solved.

another problem, when only modify the content in the form.TextValue, and then click "Save&Apply", the page says "There are no changes to apply", how to solve this problem, the code is as follows.

```javascript
o = s.taboption("custom", form.TextValue, "custom_conf",
	_(""),
	_("smartdns custom settings"));

o.rows = 20;
o.cfgvalue = function (section_id) {
	return fs.trimmed('/etc/smartdns/custom.conf');
};
o.write = function (section_id, formvalue) {
	return fs.write('/etc/smartdns/custom.conf', formvalue.trim().replace(/\r\n/g, '\n') + '\n');
};
```

**Comment by jow-** [comment]
> another problem, when only modify the content in the form.TextValue, and then click "Save&Apply", the page says "There are no changes to apply"

Yeah, this is a known problem - ignore it for now, it needs to be fixed on the LuCI side.

File: applications/luci-app-smartdns/root/etc/uci-defaults/50_luci-smartdns [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
```
Comment (jow-) [inline applications/luci-app-smartdns/root/etc/uci-defaults/50_luci-smartdns]: Your OpenWrt init script for smartdns specifies `procd_add_reload_trigger smartdns` so you can drop the entire ucitrack integration. The index cache removal has been made generic in https://github.com/openwrt/luci/commit/1f8a3c04178e21f45f3908fd476ec3b3dad73c69 so you can drop the whole uci-defaults file.

File: applications/luci-app-smartdns/root/usr/share/rpcd/acl.d/luci-app-smartdns.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{
+	"luci-app-smartdns": {
+		"description": "Grant access to LuCI app smartdns",
+		"read": {
+			"file": {
+				"/etc/smartdns/*": [ "read" ],
+				"/usr/sbin/iptables": [ "exec" ],
```
Comment (jow-) [inline applications/luci-app-smartdns/root/usr/share/rpcd/acl.d/luci-app-smartdns.json]: Please change this line to
`"/usr/sbin/iptables -t nat -nL PREROUTING": [ "exec" ],`

File: applications/luci-app-smartdns/root/usr/share/rpcd/acl.d/luci-app-smartdns.json [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+				"/usr/sbin/ip6tables": [ "exec" ],
```
Comment (jow-) [inline applications/luci-app-smartdns/root/usr/share/rpcd/acl.d/luci-app-smartdns.json]: Please change this line to:
`"/usr/sbin/ip6tables -t nat -nL PREROUTING": [ "exec" ],`

File: applications/luci-app-smartdns/htdocs/luci-static/resources/view/smartdns/smartdns.js [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+	} else {
+		renderHTML += "<b><font color=red>SmartDNS - " + _("NOT RUNNING") + "</font></b></em>";
+	}
+
+	if (redirectMode === "dnsmasq-upstream") {
+		var matchLine = "127.0.0.1#" + serverPort;
+		var dnsmasqServer = uci.get_first('dhcp', 'dnsmasq', 'server') || "";
+
+		if (dnsmasqServer.indexOf(matchLine) < 0) {
+			renderHTML += "<br></br><b><font color=red>" + _("Dnsmasq Forwared To Smartdns Failure") + "</font></b>";
```
Comment (jow-) [inline applications/luci-app-smartdns/htdocs/luci-static/resources/view/smartdns/smartdns.js]: Please change `<br></br>` to `<br />` and `<font color=red>` to `<span style="color:red">` as well as `</font>` to `</span>`

File: applications/luci-app-smartdns/htdocs/luci-static/resources/view/smartdns/smartdns.js [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+		}
+	} else if (redirectMode === "redirect") {
+		var redirectRules = (ipt || '').split(/\n/).filter(function (rule) {
+			return rule.match(/REDIRECT/) && rule.match(/dpt:53/) && rule.match("ports " + serverPort);
+		});
+
+		if (redirectRules.length <= 0) {
+			renderHTML += "<br></br><b><font color=red>" + _("IPV4 53 Port Redirect Failure") + "</font></b>";
```
Comment (jow-) [inline applications/luci-app-smartdns/htdocs/luci-static/resources/view/smartdns/smartdns.js]: Please change `<br></br>` to `<br />` and `<font color=red>` to `<span style="color:red">` as well as `</font>` to `</span>`

**Comment by pymumu** [comment]
@jow- 
All changes are done and tested, please review.

**Comment by feckert** [comment]
@jow The way I see it, all changes have been implemented.
So that the pullrequest doesn't rot I will merge them now.
And give them a try.

@pymumu thanks for your contribution

**Comment by jmarcet** [comment]
I get the following error when trying this module:

Screenshot_20200703_134655

**Comment by feckert** [comment]
@jmarcet We have already an issue for that.
I will have a look on this next week if I have more time!
See https://github.com/openwrt/luci/issues/4226

But I think we have to add the uci permition read/write to the following uci file smartdns to https://github.com/openwrt/luci/blob/master/applications/luci-app-smartdns/root/usr/share/rpcd/acl.d/luci-app-smartdns.json
Please have a look for uci read/write permissions howto solve this to the luci-app-adblock.
Write:
https://github.com/openwrt/luci/blob/master/applications/luci-app-adblock/root/usr/share/rpcd/acl.d/luci-app-adblock.json#L5
Read:
https://github.com/openwrt/luci/blob/master/applications/luci-app-adblock/root/usr/share/rpcd/acl.d/luci-app-adblock.json#L28

**Pull request merged**